lxd/storage/backend_lxd.go

```diff
@@ -0,0 +1,682 @@
+			_, snapName, _ := shared.ContainerGetParentAndSnapshotName(srcSnapshot.Name)
+			newSnapshotName := fmt.Sprintf("%s%s%s", volName, shared.SnapshotDelimiter, snapName)
+
+			// Create database entry for new storage volume.
+			err = VolumeDBCreate(b.state, b.name, newSnapshotName, srcSnapshot.Description, db.StoragePoolVolumeTypeNameCustom, true, config)
+			if err != nil {
+				return err
+			}
+
+			revertDBVolumes = append(revertDBVolumes, newSnapshotName)
+		}
+	}
+
+	srcVol := srcPool.newVolume(drivers.VolumeTypeCustom, drivers.ContentTypeFS, srcVolName, nil)
+	targetVol := b.newVolume(drivers.VolumeTypeCustom, drivers.ContentTypeFS, volName, config)
+	err = b.driver.CreateVolumeFromCopy(targetVol, srcVol, copySnapshots, op)
+	if err != nil {
+		return err
+	}
+
+	revertDBVolumes = nil // Don't revert DB volumes.
+	return nil
+}
+
+// MigrateCustomVolume sends a volume for migration.
+func (b *lxdBackend) MigrateCustomVolume(conn *websocket.Conn, args migration.VolumeSourceArgs, op *operations.Operation) error {
+	vol := b.newVolume(drivers.VolumeTypeCustom, drivers.ContentTypeFS, args.Name, nil)
+	err := b.driver.MigrateCustomVolume(vol, conn, args, op)
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// CreateCustomVolumeFromMigration receives a volume being migrated.
+func (b *lxdBackend) CreateCustomVolumeFromMigration(conn *websocket.Conn, args migration.VolumeTargetArgs, op *operations.Operation) error {
+	// Create slice to record DB volumes created if revert needed later.
+	revertDBVolumes := []string{}
+	defer func() {
+		// Remove any DB volume rows created if we are reverting.
+		for _, volName := range revertDBVolumes {
+			b.state.Cluster.StoragePoolVolumeDelete("default", volName, db.StoragePoolVolumeTypeCustom, b.ID())
+		}
+	}()
+
+	// Create database entry for new storage volume.
+	err := VolumeDBCreate(b.state, b.name, args.Name, args.Description, db.StoragePoolVolumeTypeNameCustom, false, args.Config)
+	if err != nil {
+		return err
+	}
+
+	revertDBVolumes = append(revertDBVolumes, args.Name)
+
+	if len(args.Snapshots) > 0 {
+		for _, snapName := range args.Snapshots {
+			newSnapshotName := fmt.Sprintf("%s%s%s", args.Name, shared.SnapshotDelimiter, snapName)
+
+			// Create database entry for new storage volume snapshot.
+			err = VolumeDBCreate(b.state, b.name, newSnapshotName, args.Description, db.StoragePoolVolumeTypeNameCustom, true, args.Config)
+			if err != nil {
+				return err
+			}
+
+			revertDBVolumes = append(revertDBVolumes, newSnapshotName)
+		}
+	}
+
+	vol := b.newVolume(drivers.VolumeTypeCustom, drivers.ContentTypeFS, args.Name, args.Config)
+	err = b.driver.CreateVolumeFromMigration(vol, conn, args, op)
+	if err != nil {
+		return nil
+	}
+
+	revertDBVolumes = nil
+	return nil
+}
+
+// RenameCustomVolume renames a custom volume and its snapshots.
+func (b *lxdBackend) RenameCustomVolume(volName string, newVolName string, op *operations.Operation) error {
+	if shared.IsSnapshot(volName) {
```

Comment:
So I think it'd be good to decide who's responsibility it is to validate this stuff, whether it's on the backend (seems easier) or on the driver, but duplicating all those checks just makes the code longer with no clear benefit.

Reply:
@stgraber yeah happy to do this just in backend.